File: JavaScript/__tests__/roman_numerals-test.js
/*
 * Test Suite for Roman Numeral Converter
 * Copyright © MMXVI-MMXVIII, Chris Warrick.
 * All rights reserved.
 * License: 3-clause BSD, see main file for details.
 */

test("edge cases, toRoman", () => {
    const romanNumerals = require("../roman_numerals");
    expect(romanNumerals.toRoman(1)).toBe("I");
    expect(romanNumerals.toRoman(2)).toBe("II");
    expect(romanNumerals.toRoman(3)).toBe("III");
    expect(romanNumerals.toRoman(4)).toBe("IV");
    expect(romanNumerals.toRoman(5)).toBe("V");
    expect(romanNumerals.toRoman(6)).toBe("VI");
    expect(romanNumerals.toRoman(1234)).toBe("MCCXXXIV");
    expect(romanNumerals.toRoman(1958)).toBe("MCMLVIII");
    expect(romanNumerals.toRoman(2222)).toBe("MMCCXXII");
    expect(romanNumerals.toRoman(3999)).toBe("MMMCMXCIX");
});

test("invalid numbers, toRoman", () => {
    const romanNumerals = require("../roman_numerals");
    expect(romanNumerals.toRoman(0)).toBe(-1);
    expect(romanNumerals.toRoman(-1)).toBe(-1);
    expect(romanNumerals.toRoman(4000)).toBe(-1);
});

test("all numbers, toRoman", () => {
    const romanNumerals = require("../roman_numerals");
    const fs = require("fs");
    var file = fs.readFileSync("../test_data.txt", {"encoding": "utf-8"});
    var lines = file.split("\n");
    for (var i = 0; i < lines.length; i++) {
        var t = lines[i].split(" ");
        var integer = parseInt(t[0]);
        var roman = t[1];
        if (typeof roman !== "undefined" && typeof integer !== "undefined") {
            expect(romanNumerals.toRoman(integer)).toBe(roman);
        }
    }
});

test("edge cases, fromRoman", () => {
    const romanNumerals = require("../roman_numerals");
    expect(romanNumerals.fromRoman("I")).toBe(1);
    expect(romanNumerals.fromRoman("Ii")).toBe(2);
    expect(romanNumerals.fromRoman("iii")).toBe(3);
    expect(romanNumerals.fromRoman("  iv  ")).toBe(4);
    expect(romanNumerals.fromRoman("V")).toBe(5);
    expect(romanNumerals.fromRoman("ViIi")).toBe(8);
});

test("all numbers, fromRoman", () => {
    const romanNumerals = require("../roman_numerals");
    const fs = require("fs");
    var file = fs.readFileSync("../test_data.txt", {"encoding": "utf-8"});
    var lines = file.split("\n");
    for (var i = 0; i < lines.length; i++) {
        var t = lines[i].split(" ");
        var integer = parseInt(t[0]);
        var roman = t[1];
        if (typeof roman !== "undefined" && typeof integer !== "undefined") {
            expect(romanNumerals.fromRoman(roman)).toBe(integer);
        }
    }
});

File: Java/com/chriswarrick/roman_numerals/RomanNumeralsTest.java
/*
 * Test Suite for Roman Numeral Converter
 * Copyright © MMXVI-MMXVII, Chris Warrick.
 * All rights reserved.
 * License: 3-clause BSD, see main file for details.
 */

package com.chriswarrick.roman_numerals;

import static org.junit.Assert.*;
import org.junit.Test;
import java.io.FileReader;
import java.util.Scanner;

public class RomanNumeralsTest {

    @Test
    public void testToRoman() throws Exception {
        assertEquals("I", RomanNumerals.toRoman(1));
        assertEquals("II", RomanNumerals.toRoman(2));
        assertEquals("III", RomanNumerals.toRoman(3));
        assertEquals("IV", RomanNumerals.toRoman(4));
        assertEquals("V", RomanNumerals.toRoman(5));
        assertEquals("VI", RomanNumerals.toRoman(6));
        assertEquals("MCCXXXIV", RomanNumerals.toRoman(1234));
        assertEquals("MCMLVIII", RomanNumerals.toRoman(1958));
        assertEquals("MMCCXXII", RomanNumerals.toRoman(2222));
        assertEquals("MMMCMXCIX", RomanNumerals.toRoman(3999));
    }

    @Test
    public void testToRomanInvalid() throws Exception {
        try {
            RomanNumerals.toRoman(0);
            fail("0 does not throw exception");
        } catch (Exception e) {
            assertEquals("Integer must be between 1 and 3999", e.getMessage());
        }

        try {
            RomanNumerals.toRoman(-1);
            fail("-1 does not throw exception");
        } catch (Exception e) {
            assertEquals("Integer must be between 1 and 3999", e.getMessage());
        }

        try {
            RomanNumerals.toRoman(4000);
            fail("4000 does not throw exception");
        } catch (Exception e) {
            assertEquals("Integer must be between 1 and 3999", e.getMessage());
        }
    }

    @Test
    public void testToRomanFromFile() throws Exception {
        FileReader fr = new FileReader("../test_data.txt");
        Scanner s = new Scanner(fr);
        int integer;
        String roman;
        while (s.hasNext()) {
            integer = s.nextInt();
            roman = s.next();
            assertEquals(roman, RomanNumerals.toRoman(integer));
        }
        fr.close();
        s.close();
    }

    @Test
    public void testFromRomanFromFile() throws Exception {
        FileReader fr = new FileReader("../test_data.txt");
        Scanner s = new Scanner(fr);
        int integer;
        String roman;
        while (s.hasNext()) {
            integer = s.nextInt();
            roman = s.next();
            assertEquals(integer, RomanNumerals.fromRoman(roman));
        }
        fr.close();
        s.close();
    }
}
